lib/src/scanner.dart

```diff
@@ -1587,10 +1587,24 @@ class Scanner {
         _inBlockContext,
       SP || TAB || LF || CR || BOM => false,
       NEL => true,
-      _ => _isStandardCharacter(char)
+      _ => _isStandardCharacterAt(offset)
     };
   }
 
+  bool _isStandardCharacterAt(int offset) {
+    var first = _scanner.peekChar(offset);
+    if (first == null) return false;
+
+    if (isHighSurrogate(first)) {
+      var next = _scanner.peekChar(offset + 1);
+      if (next != null && isLowSurrogate(next)) {
+        return _isStandardCharacter(decodeSurrogatePair(first, next));
```

Comment:
There are no non-standard characters in the astral planes, so you can just return `true` here, without needing to decode.

It worries me a little to have a one-or-two code unit lookahead with a `bool` result, because returning `true` doesn't tell the reader how many code units it takes up. That means we're going to do the *same* surrogate pair matching again later. That seems wasteful. 
(Do we have any kind of performance measuring on this parser? It doesn't seem particularly efficient, with a pattern of first checking something to be valid, then parsing or consuming it, instead of doing it in one step.)

Reply:
If we don't care about syncing the definition of "characters" with the yaml spec, the simplest solution would be foolow [another pull request](https://github.com/dart-lang/yaml/pull/125/commits/2532543315a10d39702bb559837aa90b10e5f6e6), i.e. just add the surrogate areas to the allowed list, maybe a little bit stricter to something like:

```dart
  bool _isStandardCharacter(int char) =>
      (char >= 0x020 && char <= 0x007E) ||
      (char >= 0x00A0 && char <= 0xD7FF) ||
      (char >= 0xE000 && char <= 0xFFFD) ||
      (char >= 0xD800 && char <= 0xDFFF) || // add this
      (char >= 0x10000 && char <= 0x10FFFF); // and remove this check as it will never be true
```

Then there will be no need to use `_scanner.readCodePoint();`. And the performance should just be the same.

Reply:
@jonasfj - do you have an opinion on removing complexity by not fully following the yaml spec?

Reply:
Personally, I'm torn on that. No text should be containing unpaired surrogates, so 99.999% of the time, it makes not difference whether they're accepted, and it will just be faster. But I also like following specifications strictly!

We really should be parsing directly from UTF-8 if we can. Most of the time, input is a UTF-8 file, and decoding UTF-8 to UTF-16 first is just wasteful overhead, which *also* introduces the possibility of unpaired surrogates. Correctly decoded UTF-8 cannot contain unpaired surrogates (WTF-8 can, so we should just not accept that for YAML). 
We can then decode the slices that are string values and leave the rest as as-is (and not decode comments at all).

(Which is also another way to say that my attempts at minuscule optimizations are completely irrelevant if we are that inefficient from the very start.)

Reply:
After reading your comment, I’m leaning more toward a simpler approach, i.e. just tweak `_isStandardCharacter()` and allow the UTF-16 surrogate range to pass through.
- Currently, the library also doesn’t check for invalid (unpaired) UTF-16 surrogates in quoted strings.
- Relaxing the check in `_isStandardCharacter()` may not improve but at least won’t break the expectation of standard compliance. In short it won't make things worse.
- Talking about expectation, I also think it is better for a library to behave the same when parsing quoted and unquoted string. While unlikely, just doing surrogate pair validation on unquoted string might introduce another report of unexpected behavior between the two type of strings.
- Less to no performance impact compared to this one.

@lrhn's suggestion sound valuable, but will take longer to implement (to me it sounds like a change from the ground up).

Anyway, if I want to submit a change towards `_isStandardCharacter()`, is it better to open a new pull request (#159 probably needs a rebase and add some test cases)? Or a force push this pull request is preferred? Thank you!

Reply:
Oops, looks like what I said above was wrong: The Yaml spec says:
> To ensure [JSON compatibility](https://yaml.org/spec/1.2.2/#yaml-directives), YAML [processors](https://yaml.org/spec/1.2.2/#processes-and-models) must allow all non-C0 characters inside [quoted scalars](https://yaml.org/spec/1.2.2/#double-quoted-style). To ensure readability, non-printable characters should be [escaped](https://yaml.org/spec/1.2.2/#escaped-characters) on output, even inside such [scalars](https://yaml.org/spec/1.2.2/#scalars).

So quoted and unquoted scalers are not analogous in this context. Sorry for the noise!